### ISSUE TITLE
Add force feed back for shifter

### ISSUE THREAD

**Issue by hoantv** [hoantv/LCPedal#2]
- Use x,y shifter 
- play ffb when clutch is not reach a certain value

**Comment by huseyinctlbs** [comment]
Hi. I'm closely interested in your project. I'm so glad to see you have made some serious progress on something I've been researching for a long time. I'm especially curious about the ffb effect you plan to make for the xy shifter.

I would like to do a 3d printed xy shifter to test with your software if it's appropriate for you.(i will share all the desings with you).

Thanks a lot for sharing those information with us.

**Comment by hoantv** [comment]
I am investgating 2 options
1. Use frex prototype with button. I sent my idea to one guy on DIY Sim Racer Group FB. He said he will draw 2d, 3d for me. But i dont know when his finish.
2. I used real shifter and 2 encoder (or rotation potentiometer) as X,Y axis. 2 cable of real shifter control encoder rotation (X cable, Y cable).
A motor connect to X cable to control ffb. The motor can be mg996r (or maybe use AC servo, so can its encoder as X encoder). 
I prefer to 2nd option with ac servo.  it’s too expensive but easy to do.

**Comment by hoantv** [comment]
I am investgating 2 options

- Use frex prototype with button. I sent my idea to one guy on DIY Sim Racer Group FB. He said he will draw 2d, 3d for me. But i dont know when his finish.
- Use real shifter and 2 encoder (or rotation potentiometer) as X,Y axis. 2 cable of real shifter control encoder rotation (X cable, Y cable).
A motor connect to X cable to control ffb. The motor can be mg996r (or maybe use AC servo, so can its encoder as X encoder).
I prefer to 2nd option with ac servo. it’s too expensive but easy to do.

**Comment by huseyinctlbs** [comment]
I think the second way would be more expensive my friend. Arthur Mellen Wellington said that about the engineering "An engineer can do for a dollar what any fool can do for two is attributed to him" :))

But some persons from endonesia (maybe you have seen them) have made xy shifter with just 2 pot using with arduino instead of button prototype. This could be more easier and more cheaper than using real car mechanism :)) and maybe can use a vibration motor that vibrates the shifter stick. 

Or can be there will be a geared system on each axis and a servo or dc motor that block the shifter stick if not pressed on clutch pedal. 

I think these methods are more appropriate to the words of "Wellington" :))

**Comment by hoantv** [comment]
i see. I though about use joystick as shifter before :). But i dont know about mechanic, so i have to choose some available materials. 
At the moment, all are ideas :d.

**Comment by hoantv** [comment]
i see. I though about use joystick as shifter before :). I saw it is too weak, maybe broken.
I though about 2 poteniometer. But i dont know about mechanic, so i have to choose some available materials.
At the moment, all are ideas :d.

**Comment by hoantv** [comment]
It is seem that you are making new firmware for ffb wheel?

**Comment by huseyinctlbs** [comment]
Oh yes. I have done a shifter from ffb joystick with a software. It was setting the limits of the gears with a template and hold the stick at gear positions. But i remember that it has not have the clutch option for blocking the shifter stick. 

Anyway these are just ideas for now like you said. And i just want to tell you that these ideas can be made with software. If you need some help, you can ask me.  lazy_1989@hotmail.com is my mail.

Im waiting to hear good news from you. Thanks a lot for sharing such a project with people.

**Comment by huseyinctlbs** [comment]
Sorry i haven't seen your post. Yes i'm trying to do that but i don't have a startpoint for it. Then your github will be a good startpoint to me. Let's be in contact. :))
Thanks again.

**Comment by hoantv** [comment]
If you are interested in, Pls join my projects.
I will make other version for other MCU (maybe stm32).

**Comment by huseyinctlbs** [comment]
Yes i have joined. As far as I know, STM32 series development boards are more powerful than arduinos. But I guess the arduino should come after STM. Stm series boards requires more experience and the necessary experience could be learned in the arduinos. Do you have another page? Facebook or instagram that we can connect?

**Comment by hoantv** [comment]
//XY shifter use potentiometer -> maybe use encoder

#define xRange 1200
#define yRange 900

#define xPin A0
#define yPin A1

uint8_t xPosition = 1;
uint8_t yPosition = 1;
uint8_t shifterPosition = 0;
int xValue = 0, yValue = 0;

// calculate Shifter Position  
  xValue = analogRead(A0);
  yValue = analogRead(A1); //Value Map to PMW signal.
  xPosition = map(xValue, 0, xRange, 0, 3);
  yPosition = map(yValue, 0, yRange, 0, 2);
  switch (xPosition) {
    case 0:
      switch (yPosition) {
        case 0:
          shifterPosition = 1;
          break;
        case 2:
          shifterPosition = 2;
        default:
          shifterPosition = 0;
          break;
      }
      break;
    case 1:
      switch (yPosition) {
        case 0:
          shifterPosition = 3;
          break;
        case 2:
          shifterPosition = 4;
        default:
          shifterPosition = 0;
          break;
      }
      break;
    case 2:
      switch (yPosition) {
        case 0:
          shifterPosition = 5;
          break;
        case 2:
          shifterPosition = 6;
        default:
          shifterPosition = 0;
          break;
      }
      break;
    case 3:
      switch (yPosition) {
        case 0:
          shifterPosition = 7;
          break;
        case 2:
          shifterPosition = 8; //Reverse
        default:
          shifterPosition = 0;
          break;
      }
      break;
    default:
      break;
  }

**Comment by hoantv** [comment]
This is my idea about ffb XY shifter:

- M is a motor or servo (i doesn't know which is enough strong, may be i choose servo mg996r or motor 775/ BTS7960)
- P can be potentiometer or encoder for Y axis (i did draw X axis here, because it doesn't join ffb process)
- S is shifter shaft. It is horizontal (if using frex prototype or real shifter with cables) or vertical (if using H partern shifter). 
- L is sping ball latch (i doesn't know exaclty in english), it can be in 1 (like frex) or 2 position (like fanatec).
Working:
- When clutch is not greater thresshold (95% for example), when you move shifter up/down, it change value of Y axis. 
- if Y value is in range (y1,y2; (y1+y2)/2 is nature position 0). The motor will shake by pwm signal from MCU (arduino).
- if Y value is out of range (y1,y2). beside shaking, the motor will rotate reverse to prevent Y out of range (y1,y2).

**Comment by hoantv** [comment]
This is my idea about ffb XY shifter:

- M is a motor or servo (i doesn't know which is enough strong, may be i choose servo mg996r or motor 775/ BTS7960)
- P can be potentiometer or encoder for Y axis (i did draw X axis here, because it doesn't join ffb process)
- S is shifter shaft. It is horizontal (if using frex prototype or real shifter with cables) or vertical (if using H partern shifter). 
- L is sping ball latch (i doesn't know exaclty in english), it can be in 1 (like frex) or 2 position (like fanatec).
Working:
- When clutch is not greater thresshold (95% for example), when you move shifter up/down, it change value of Y axis. 
- if Y value is in range (y1,y2; (y1+y2)/2 is nature position 0). The motor will shake by pwm signal from MCU (arduino).
- if Y value is out of range (y1,y2). beside shaking, the motor will rotate reverse to prevent Y out of range (y1,y2).
